File: Leetcode/C++/101. Symmetric Tree/solution.cpp
// https://leetcode.com/problems/symmetric-tree/description/

/**
 * Definition for a binary tree node. */
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

#include <queue>
#include <vector>

using namespace std;

class Solution
{
public:
    bool isSymmetric(TreeNode *root)
    {
        if (root == nullptr)
        {
            return true;
        }
        TreeNode *left, *right;
        queue<TreeNode *> q1, q2;
        q1.push(root->left);
        q2.push(root->right);
        while (!q1.empty())
        {
            left = q1.front();
            right = q2.front();
            q1.pop();
            q2.pop();
            if (left == nullptr && right == nullptr)
            {
                continue;
            }
            else if ((left == nullptr || right == nullptr) || left->val != right->val)
            {
                return false;
            }
            q1.push(left->left);
            q2.push(right->right);

            q1.push(left->right);
            q2.push(right->left);
        }
        return true;
    }
};

File: Leetcode/Python/101. Symmetric Tree/solution.py
# https://leetcode.com/problems/symmetric-tree/description/


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:

        def symmetric(left, right):
            if left and right:  # Filter (left and right) for (left or right)
                pass
            elif left or right:
                return False
            else:   # Both None
                return True

            if left.val != right.val:
                return False

            return symmetric(left.left, right.right) and symmetric(left.right, right.left)
        return symmetric(root.left, root.right)
